engine/access/rpc/backend/backend_scripts.go

```diff
@@ -108,17 +110,53 @@ func (b *backendScripts) executeScriptOnExecutor(
 	script []byte,
 	arguments [][]byte,
 ) ([]byte, error) {
-	// find few execution nodes which have executed the block earlier and provided an execution receipt for it
-	scriptExecutors, err := b.findScriptExecutors(ctx, blockID)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to find script executors at blockId %v: %v", blockID.String(), err)
-	}
 	// encode to MD5 as low compute/memory lookup key
 	// CAUTION: cryptographically insecure md5 is used here, but only to de-duplicate logs.
 	// *DO NOT* use this hash for any protocol-related or cryptographic functions.
 	insecureScriptHash := md5.Sum(script) //nolint:gosec
-
 	// try execution on Archive nodes first
+	archiveResult, err := b.executeScriptOnAvailableArchiveNodes(ctx, blockID, script, arguments, insecureScriptHash)
+	// try execution nodes if the script wasn't executed
+	if b.scriptExecValidation {
+		execNodeResult, errExec := b.executeScriptOnAvailableExecutionNodes(
+			ctx, blockID, script, arguments, insecureScriptHash)
+		if bytes.Equal(execNodeResult, archiveResult) && errExec == err {
+			b.logScriptExecutionComparison(blockID, insecureScriptHash,
+				"script execution results on Archive node and EN are equal")
+		} else {
+			b.logScriptExecutionComparison(blockID, insecureScriptHash,
+				"script execution results on Archive node and EN are not equal")
+		}
+		// return EN results by default
+		return execNodeResult, errExec
+	} else {
+		// execute on execution nodes if it's not a script error
+		if err != nil && status.Code(err) != codes.InvalidArgument {
```

Comment:
Should we just always re-execute if there's an error? It looks like the current implementation on the archive node will return `InvalidArgument` for all invoker errors:
https://github.com/onflow/flow-archive/blob/a3113fce4d808b59b18841307259e234265ddc96/service/access/server.go#L446-L457

In theory that always means there's an actual script execution error (instead of block/data), but since this is early testing it probably makes sense to be more conservative.

Reply:
yeah that was the thinking, stay conservative if there's any chance the EN would be able to execute it

Reply:
how about we always reexecute on error then?

Reply:
then that won't be any different from the "validation mode", apart from checking the results. we want to fallback to ENs only if the register data / RN is unavailable which is checked before the invoker. It gives us errors like `NotFound`

Reply:
I don't want to hold this up, so I approved. My point is that this is new and the errors returned may not be exactly what we expect. We already ran into that with the earlier FVM bug. Reexecuting on any error would allow us to collect the metrics and logs with lower risk of impacting user traffic.

Reply:
I agree with the risk mitigation idea, which is why we only intend to run it on "validation-mode" for now